waiter/src/waiter/process_request.clj

```diff
@@ -873,20 +875,41 @@
             (do
               (log/info "x-waiter-maintenance is not supported as an on-the-fly header"
                         {:service-description service-description-template :token token})
-              (-> {:message "The maintenance parameter is not supported for on-the-fly requests"
-                   :status http-400-bad-request
-                   :waiter-headers waiter-headers}
-                  (utils/data->error-response request)))
+              (on-header-error {:message "The maintenance parameter is not supported for on-the-fly requests"
+                                :status http-400-bad-request
+                                :waiter-headers waiter-headers}
+                               request))
             (some? maintenance)
             (do
               (log/info (str "token " token " is in maintenance mode"))
               (meters/mark! (metrics/waiter-meter "maintenance" "response-rate"))
-              (-> {:details response-map
-                   :message (get maintenance "message")
-                   :status http-503-service-unavailable}
-                  (utils/data->maintenance-mode-response request)))
+              (on-maintenance-mode-error {:details response-map
+                                          :message (get maintenance "message")
+                                          :status http-503-service-unavailable}
+                                         request))
             :else (handler request)))))
 
+(defn wrap-maintenance-mode
+  "Middleware to check for maintenance mode of a token and if improper maintenance mode header is provided"
+  [handler]
+  (make-maintenance-mode
+    handler
+    (fn [data-map request]
+      (-> data-map
+          (utils/data->error-response request)))
```

Comment:
Nitpick: we can avoid using the arrow operator here as there is a single invocation.